iree/compiler/Dialect/Flow/Transforms/Passes.cpp

```diff
@@ -65,17 +65,18 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager) {
   //----------------------------------------------------------------------------
   passManager.addPass(createCanonicalizerPass());
 
+  // Flatten structured control flow to our CFG.
+  passManager.addNestedPass<FuncOp>(mhlo::createLegalizeControlFlowPass());
+  passManager.addNestedPass<FuncOp>(createHLOPreprocessingPass());
+
   // Frontload linalg-on-tensors transformations and dispatch region creation.
   if (clEnableLinalgOnTensorsDispatch) {
+    passManager.addNestedPass<FuncOp>(createCanonicalizerPass());
     addHLOToLinalgOnTensorsPasses(passManager);
     passManager.addNestedPass<FuncOp>(createDispatchLinalgOnTensorsPass(
         clLinalgOnTensorsTileSizes, clLinalgOnTensorsEnableFusion));
   }
 
-  // Flatten structured control flow to our CFG.
-  passManager.addNestedPass<FuncOp>(mhlo::createLegalizeControlFlowPass());
-  passManager.addNestedPass<FuncOp>(createHLOPreprocessingPass());
-
   // Convert TOSA ops to Linalg-on-tensor ops.
   passManager.addNestedPass<FuncOp>(tosa::createTosaToLinalgOnTensors());
```

Comment:
so we aren't playing whackamole here we need to fix this up properly asap - for now, can you at least move the passes that turn things into linalg above where we require things to be in linalg? (it was a nasty hack that it was added in that the top - it never belonged there)